azurelinuxagent/common/protocol/wire.py

```diff
@@ -913,49 +913,51 @@ def _call_hostplugin_with_container_check(self, host_func):
 
         return ret
 
-    def send_request_using_appropriate_channel(self, direct_func, host_func):
-        # A wrapper method for all function calls that send HTTP requests. The purpose of the method is to
-        # define which channel to use, direct or through the host plugin. For the host plugin channel,
-        # also implement a retry mechanism.
-
-        # By default, the direct channel is the default channel. If that is the case, try getting a response
-        # through that channel. On failure, fall back to the host plugin channel.
+    def __send_request_using_host_channel(self, host_func):
+        ret = self.__call_hostplugin_with_container_check(host_func)
 
-        # When using the host plugin channel, regardless if it's set as default or not, try sending the request first.
-        # On specific failures that indicate a stale goal state (such as resource gone or invalid container parameter),
-        # refresh the goal state and try again. If successful, set the host plugin channel as default. If failed,
-        # raise the exception.
+        if ret and not HostPluginProtocol.is_default_channel():
+            # Set host plugin channel as default if it succeeded.
+            logger.info("Switching default channel to host plugin.")
+            HostPluginProtocol.set_default_channel(True)
 
-        # NOTE: direct_func and host_func are passed as lambdas. Be careful about capturing goal state data in them as
-        # they will not be refreshed even if a goal state refresh is called before retrying the host_func.
+        return ret
 
-        if not HostPluginProtocol.is_default_channel():
-            ret = None
-            try:
-                ret = direct_func()
+    def __send_request_using_direct_channel(self, direct_func):
+        ret = None
+        try:
+            ret = direct_func()
 
-                # Different direct channel functions report failure in different ways: by returning None, False,
-                # or raising ResourceGone or InvalidContainer exceptions.
-                if not ret:
-                    logger.periodic_info(logger.EVERY_HOUR, "[PERIODIC] Request failed using the direct channel, "
-                                                            "switching to host plugin.")
-            except (ResourceGoneError, InvalidContainerError) as e: # pylint: disable=C0103
+            # Different direct channel functions report failure in different ways: by returning None, False,
+            # or raising ResourceGone or InvalidContainer exceptions.
+            if not ret:
                 logger.periodic_info(logger.EVERY_HOUR, "[PERIODIC] Request failed using the direct channel, "
-                                                        "switching to host plugin. Error: {0}".format(ustr(e)))
+                                                        "trying host plugin next.")
+        except (ResourceGoneError, InvalidContainerError) as e:  # pylint: disable=C0103
```

Comment:
those are errors returned by the host ga plugin. why are we checking for those on the direct channel?

Reply:
This is existing behaviour. I can't remember right now why direct channel also has this check, I can look at the history... but I am hesitant to change it.

Reply:
please look at the history... we are trying to improve this code

Reply:
depending on your investigation we may decide to leave as is with documentation, change it in this PR or change it later

Reply:
From what I remember `InvalidContainerError` was raised by HostGA when the containerId is out of sync. I think `ResourceGoneError` was also due to something with HostGA but its definitely better to confirm (maybe @jacwil  can also shed some light here)

Reply:
@larohra yes, those are raised by HostGA plugin. It is odd that we using them for the direct channel.

Although those checks were not introduced by this PR, we should look at the code history to double-check we did not copy-paste those checks by mistake when refactoring this code previously. If they have been there from the beginning we should mark this code or open a task/issue to correct this later (hopefully soon)

Reply:
These errors are specific to the host channel. We don't need to check for them on the direct channel, I updated the code.